File: openacc-wg-2016/matmul_omp.c
#include <stdio.h>
#include "ticktock.h"

#define SIZE 2250

int a[SIZE][SIZE];
int b[SIZE][SIZE];
int c[SIZE][SIZE];

int main(){
    int i,j,k;
    int sum = 0;

    for(i=0;i<SIZE;i++)
        for(j=0;j<SIZE;j++){
            a[i][j] = (i%50);
            b[i][j] = (j%30);
            c[i][j] = 0;
        }
    tick_tock tt;
    tick(&tt); 
    printf("b[523][789] = %d\n",b[523][789]);
    for(i=0;i<SIZE;i++)
        #pragma omp parallel for private(k,sum)
        for(j=0;j<SIZE;j++){
            sum = 0;
            for(k=0;k<SIZE;k++)
                sum += a[i][k] * b[k][j];
            c[i][j] = sum;
        }
    tock(&tt);    
    printf("c[523][789] = %d\n",c[523][789]);
    return 0;     
}

File: openacc-wg-2016/matmul_omp.f90
program matmul_acc
    
    implicit none
    integer,parameter :: SIZE=2250

    integer :: a(SIZE,SIZE), b(SIZE,SIZE), c(SIZE,SIZE);
    integer :: sum
    integer :: i,j,k
    integer :: it0,it1

    do i=1,SIZE
        do j=1,SIZE
            a(i,j) = mod(i,50);
            b(i,j) = mod(j,30);
            c(i,j) = 0;
        end do
    end do
    write(*,"(A,1I14)") "b(523,789) = ",b(523,789)
    call tick(it0,it1) 
    do i=1,SIZE
        !$omp parallel do private(k,sum)
        do j=1,SIZE
            sum = 0
            do k=1,SIZE
                sum = sum + a(i,k) * b(k,j)
            end do
            c(i,j) = sum
        end do
        !$omp end parallel do
    end do
    call tock(it0,it1)    
    write(*,"(A,1I14)") "c(523,789) = ",c(523,789)
end program matmul_acc
